File: java-string-calculator/src/main/java/io/github/shirohoo/stringcalculator/ArithmeticOperator.java
package io.github.shirohoo.stringcalculator;

import static java.util.Arrays.stream;
import java.util.function.DoubleBinaryOperator;

public enum ArithmeticOperator {
    PLUS('+', (left, right) -> left + right),
    MINUS('-', (left, right) -> left - right),
    MULTIPLY('*', (left, right) -> left * right),
    DIVIDE('/', (left, right) -> left / right),
    ;

    private final char operator;
    private final DoubleBinaryOperator doubleBinaryOperator;

    ArithmeticOperator(char operator, DoubleBinaryOperator doubleBinaryOperator) {
        this.operator = operator;
        this.doubleBinaryOperator = doubleBinaryOperator;
    }

    public static ArithmeticOperator findByOperator(char operator) {
        return stream(values())
                .filter(it -> it.operator == operator)
                .findFirst()
                .orElseThrow();
    }

    public double apply(double left, double right) {
        if (this == DIVIDE && right == 0) {
            throw new ArithmeticException("not divisible by zero");
        }
        return doubleBinaryOperator.applyAsDouble(left, right);
    }
}

File: kt-string-calculator/src/main/kotlin/io/github/shirohoo/stringcalculator/ArithmeticOperator.kt
package io.github.shirohoo.stringcalculator

enum class ArithmeticOperator(
    private val operator: Char,
    private val doubleBinaryOperator: (Double, Double) -> Double
) {
    PLUS('+', { left: Double, right: Double -> left + right }),
    MINUS('-', { left: Double, right: Double -> left - right }),
    MULTIPLY('*', { left: Double, right: Double -> left * right }),
    DIVIDE('/', { left: Double, right: Double -> left / right })
    ;

    fun apply(left: Double, right: Double): Double {
        if (this == DIVIDE && right == 0.0) {
            throw ArithmeticException("not divisible by zero")
        }
        return doubleBinaryOperator(left, right)
    }

    companion object {
        fun findByOperator(operator: Char): ArithmeticOperator {
            return values().first { it.operator == operator }
        }
    }
}
